src/rockstor/system/directory_services.py

```diff
@@ -0,0 +1,332 @@
+"""
+Copyright (c) 2012-2020 RockStor, Inc. <http://rockstor.com>
+This file is part of RockStor.
+
+RockStor is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published
+by the Free Software Foundation; either version 2 of the License,
+or (at your option) any later version.
+
+RockStor is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+"""
+
+import os
+import stat
+
+import re
+
+from osi import append_to_line, run_command
+from tempfile import mkstemp
+from shutil import move
+import logging
+
+from system.services import systemctl
+
+logger = logging.getLogger(__name__)
+
+NSSWITCH_FILE = "/etc/nsswitch.conf"
+SSSD_FILE = "/etc/sssd/sssd.conf"
+NET = "/usr/bin/net"
+REALM = "/usr/sbin/realm"
+ADCLI = "/usr/sbin/adcli"
+OPENSSL = "/usr/bin/openssl"
+
+
+def validate_tls_cert(server, cert):
+    """
+    Run openssl s_client -connect server:389 -CAfile path-to-cert
+    to verify the provided TLS certificate against the LDAP server.
+    :param server: String - FQDN of the LDAP server
+    :param cert: String - Absolute path to the TLS certificate
+    :return:
+    """
+    cmd = [
+        OPENSSL,
+        "s_client",
+        "-connect",
+        "{}:389".format(server),
+        "-CAfile",
+        cert,
+    ]
+    o, e, rc = run_command(cmd, throw=False)
+    if "Verification: OK" not in o:
+        err_msg = (
+            "Failed to validate the TLS certificate ({}).\n"
+            "out: {} err: {} rc: {}".format(cert, o, e, rc)
+        )
+        if any("fopen:No such file or directory" in err for err in e):
+            err_msg = (
+                "The TLS certificate file could not be found at {}.\n"
+                "out: {} err: {} rc: {}".format(cert, o, e, rc)
+            )
+        raise Exception(err_msg)
+
+
+def update_nss(databases, provider, remove=False):
+    """
```

Comment:
We need to take care with how this nsswitch file editor interacts with the one we already have:
config_nis() in src/rockstor/system/nis.py
that was recently updated in:
NIS regression re Rockstor 4 release candidates. Fixes #2216 #2221
And or to place them both in a common file or the like.
All good thought and this one looks to be neater for adding and removing content in it's specific way.

Reply:
Yes, I noticed that but didn't want to change the config you just implemented... most importantly, I wouldn't have been sure of how to test it functionally so I decided not to mess it with. I can create a dedicated issue so that we keep track of this and do merge & harmonize the two approaches, if you'd like.

Reply:
OK that makes sense. An issue might be good and I hadn't thought of the unit testing on these. So yes keeping them discrete may well help simplify any later testing. But it would be nice to have them in the same place at least. Or at bare minimum have their comments reference each other it they are where they belong / make most sense already.